File: core-ts/src/oracle_cli.ts
/**
 * oracle_cli — persistent oracle subprocess for differential fuzzing.
 *
 * Reads hex-encoded wire-byte strings from stdin (one per line) and
 * emits a JSON oracle report to stdout for each. EOF terminates
 * cleanly. Mirrors the C++ `oigtl_oracle_cli` binary and the Python
 * oracle `verify` command — all three emit identical JSON shapes,
 * which is the point: the `oigtl-corpus fuzz differential` runner
 * compares them line-for-line.
 *
 * Run with (after `npm run build`):
 *
 *     node dist/oracle_cli.js
 *
 * Or during development (after `tsc -p tsconfig.json --outDir build-tests`):
 *
 *     node build-tests/src/oracle_cli.js
 */

import { createInterface } from "node:readline";
import { stdin, stdout } from "node:process";

import { fromHex } from "./runtime/byte_order.js";
import { toReport, verifyWireBytes } from "./runtime/oracle.js";
import "./messages/index.js"; // side-effect: registers all 84 message classes

function emitInvalidHex(): void {
  // Emit a report with the same shape as a real failure — keeps the
  // Python consumer simple (no special-case parsing for hex errors).
  const empty = {
    ok: false,
    type_id: "",
    device_name: "",
    version: 0,
    body_size: 0,
    ext_header_size: null,
    metadata_count: 0,
    round_trip_ok: false,
    error: "oracle_cli: invalid hex input",
  };
  stdout.write(JSON.stringify(empty) + "\n");
}

const rl = createInterface({ input: stdin, crlfDelay: Infinity });

rl.on("line", (line) => {
  const clean = line.trim();
  let bytes: Uint8Array;
  try {
    bytes = fromHex(clean);
  } catch {
    emitInvalidHex();
    return;
  }
  try {
    const result = verifyWireBytes(bytes);
    stdout.write(JSON.stringify(toReport(result)) + "\n");
  } catch (e) {
    // Unexpected exception — should never happen (the oracle swallows
    // its own errors), but surface rather than crash so a real bug
    // gets recorded in the fuzzer's disagreement log.
    const err = {
      ok: false,
      type_id: "",
      device_name: "",
      version: 0,
      body_size: 0,
      ext_header_size: null,
      metadata_count: 0,
      round_trip_ok: false,
      error: `oracle_cli: uncaught ${(e as Error).name}: ${(e as Error).message}`,
    };
    stdout.write(JSON.stringify(err) + "\n");
  }
});

rl.on("close", () => {
  // EOF on stdin — exit cleanly.
  process.exit(0);
});

File: core-py/src/oigtl/oracle_cli.py
"""Persistent oracle subprocess for the differential fuzzer.

Same protocol as ``core-cpp/src/oracle_cli.cpp`` and
``core-ts/src/oracle_cli.ts``:

- stdin: one hex-encoded wire message per line (or empty line for
  a 0-byte input).
- stdout: one JSON oracle report per line, matching the stable
  cross-language shape (``ok``, ``type_id``, ``device_name``,
  ``version``, ``body_size``, ``ext_header_size``, ``metadata_count``,
  ``round_trip_ok``, ``error``).

Runs the TYPED oracle (``oigtl.runtime.oracle.typed_verify_wire_bytes``),
so it exercises the generated message classes and the numpy /
array.array coercion paths. Toggle the stdlib fallback with
``OIGTL_NO_NUMPY=1`` in the environment before starting this
process.

Invoke via:

    python -m oigtl.oracle_cli
"""

from __future__ import annotations

import json
import sys

from oigtl.runtime.oracle import typed_verify_wire_bytes


def _report_to_dict(result) -> dict:
    header = result.header
    ext = result.extended_header
    return {
        "ok": result.ok,
        "type_id": header.type_id if header else "",
        "device_name": header.device_name if header else "",
        "version": header.version if header else 0,
        "body_size": header.body_size if header else 0,
        "ext_header_size": ext.ext_header_size if ext else None,
        "metadata_count": len(result.metadata),
        "round_trip_ok": result.round_trip_ok,
        "error": result.error,
    }


def _emit_invalid_hex() -> None:
    sys.stdout.write(json.dumps({
        "ok": False,
        "type_id": "",
        "device_name": "",
        "version": 0,
        "body_size": 0,
        "ext_header_size": None,
        "metadata_count": 0,
        "round_trip_ok": False,
        "error": "oracle_cli: invalid hex input",
    }) + "\n")


def main() -> int:
    for line in sys.stdin:
        line = line.rstrip("\r\n")
        try:
            wire = bytes.fromhex(line)
        except ValueError:
            _emit_invalid_hex()
            sys.stdout.flush()
            continue
        try:
            result = typed_verify_wire_bytes(wire)
            sys.stdout.write(json.dumps(_report_to_dict(result)) + "\n")
        except Exception as exc:
            # Any uncaught exception is itself a bug — surface it as
            # a special error so the fuzzer records a disagreement
            # instead of crashing the oracle process.
            sys.stdout.write(json.dumps({
                "ok": False,
                "type_id": "",
                "device_name": "",
                "version": 0,
                "body_size": 0,
                "ext_header_size": None,
                "metadata_count": 0,
                "round_trip_ok": False,
                "error": f"oracle_cli: uncaught {type(exc).__name__}: {exc}",
            }) + "\n")
        sys.stdout.flush()
    return 0


if __name__ == "__main__":
    sys.exit(main())
